File: app/src/main/java/com/aryan/myrecon/data/GeoIntel.kt
package com.aryan.myrecon.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.*
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.concurrent.TimeUnit

/**
 * Turns GPS coordinates into a named place.
 *
 * This is the honest substitute for visual landmark recognition, which needs a
 * paid vision API. When a photo carries GPS — and many do — resolving those
 * coordinates is *more* reliable than recognising a building from pixels: Big
 * Ben resolves to "Big Ben, Bridge Street, Westminster" with a Wikidata id and
 * the eight nearest notable places, from the numbers alone.
 *
 * The trade is stated rather than hidden: no GPS, no result. Nothing is guessed
 * from image content.
 *
 * Sources, both keyless:
 *   Nominatim (OpenStreetMap) — reverse geocoding
 *   Wikipedia REST + Action API — article summary and nearby places
 */
object GeoIntel {

    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .build()

    // Nominatim's policy requires a genuine identifying agent with contact
    // details. A browser UA is a violation and gets the address blocked.
    private const val UA = "MyRecon-Android/1.0 (https://myrecon.xyz; abuse@myrecon.xyz)"

    /**
     * At most one Nominatim request per second, enforced across coroutines.
     * Exceeding it gets the IP banned, which would break the feature for every
     * user of the app — so this is not optional politeness.
     */
    private val throttle = Mutex()
    private var lastCall = 0L
    private const val MIN_INTERVAL_MS = 1_100L

    data class Place(
        val found: Boolean,
        val name: String? = null,
        val displayName: String? = null,
        val city: String? = null,
        val state: String? = null,
        val country: String? = null,
        val countryCode: String? = null,
        val postcode: String? = null,
        val wikidata: String? = null,
        val osmUrl: String? = null,
    )

    data class Article(
        val title: String,
        val description: String?,
        val extract: String?,
        val url: String?,
        val thumbnail: String?,
    )

    data class Nearby(val title: String, val distanceM: Double, val url: String)

    data class Report(
        val latitude: Double,
        val longitude: Double,
        val mapsUrl: String,
        val place: Place,
        val article: Article?,
        val nearby: List<Nearby>,
        val confidence: Int,
        val confidenceLabel: String,
        val error: String? = null,
    )

    fun validCoords(lat: Double?, lon: Double?): Boolean {
        if (lat == null || lon == null) return false
        // Reject null island exactly: 0,0 is overwhelmingly a zeroed GPS field
        // rather than a real position in the Gulf of Guinea.
        if (lat == 0.0 && lon == 0.0) return false
        return lat in -90.0..90.0 && lon in -180.0..180.0
    }

    suspend fun investigate(lat: Double, lon: Double): Report = withContext(Dispatchers.IO) {
        if (!validCoords(lat, lon)) {
            return@withContext Report(
                lat, lon, mapsUrl(lat, lon), Place(false), null, emptyList(), 0, "none",
                error = "No usable GPS coordinates.",
            )
        }

        val place = runCatching { reverseGeocode(lat, lon) }.getOrDefault(Place(false))
        val nearby = runCatching { nearbyPlaces(lat, lon) }.getOrDefault(emptyList())

        // Prefer the article Nominatim points at; otherwise the closest notable
        // place, which is usually the same feature under its common name.
        val article = runCatching {
            place.name?.let { summary(it) } ?: nearby.firstOrNull()?.let { summary(it.title) }
        }.getOrNull()

        var score = 0
        if (place.found) score += 30
        if (place.name != null) score += 25
        if (place.wikidata != null) score += 20
        if (article != null) score += 15
        if (nearby.firstOrNull()?.distanceM?.let { it <= 50 } == true) score += 10
        score = score.coerceAtMost(100)

        Report(
            latitude = lat, longitude = lon,
            mapsUrl = mapsUrl(lat, lon),
            place = place, article = article, nearby = nearby,
            confidence = score,
            confidenceLabel = when {
                score >= 70 -> "high"
                score >= 40 -> "medium"
                else -> "low"
            },
        )
    }

    private fun mapsUrl(lat: Double, lon: Double) =
        "https://www.openstreetmap.org/?mlat=$lat&mlon=$lon#map=17/$lat/$lon"

    // ── Sources ──────────────────────────────────────────────────

    private suspend fun reverseGeocode(lat: Double, lon: Double): Place {
        val body = throttled(
            "https://nominatim.openstreetmap.org/reverse" +
                "?lat=$lat&lon=$lon&format=jsonv2&addressdetails=1&extratags=1"
        ) ?: return Place(false)

        val o = body as? JsonObject ?: return Place(false)
        if (o["error"] != null) return Place(false)
        val addr = o["address"] as? JsonObject
        val extra = o["extratags"] as? JsonObject

        fun s(obj: JsonObject?, k: String) =
            (obj?.get(k) as? JsonPrimitive)?.contentOrNull?.takeIf { it.isNotBlank() }

        val osmType = s(o, "osm_type")
        val osmId = s(o, "osm_id")

        return Place(
            found = true,
            name = s(o, "name"),
            displayName = s(o, "display_name"),
            // Nominatim varies which key holds the settlement by country, so
            // fall through the plausible ones in decreasing specificity.
            city = s(addr, "city") ?: s(addr, "town") ?: s(addr, "village")
                ?: s(addr, "municipality") ?: s(addr, "county"),
            state = s(addr, "state"),
            country = s(addr, "country"),
            countryCode = s(addr, "country_code")?.uppercase(),
            postcode = s(addr, "postcode"),
            wikidata = s(extra, "wikidata"),
            osmUrl = if (osmType != null && osmId != null)
                "https://www.openstreetmap.org/$osmType/$osmId" else null,
        )
    }

    private fun nearbyPlaces(lat: Double, lon: Double, radiusM: Int = 1000, limit: Int = 8): List<Nearby> {
        val body = plainGet(
            "https://en.wikipedia.org/w/api.php?action=query&list=geosearch" +
                "&gscoord=$lat%7C$lon&gsradius=${radiusM.coerceIn(10, 10_000)}" +
                "&gslimit=${limit.coerceIn(1, 50)}&format=json"
        ) as? JsonObject ?: return emptyList()

        val results = ((body["query"] as? JsonObject)?.get("geosearch") as? JsonArray) ?: return emptyList()
        return results.mapNotNull { el ->
            val o = el as? JsonObject ?: return@mapNotNull null
            val title = (o["title"] as? JsonPrimitive)?.contentOrNull ?: return@mapNotNull null
            val dist = (o["dist"] as? JsonPrimitive)?.contentOrNull?.toDoubleOrNull() ?: 0.0
            Nearby(title, dist, "https://en.wikipedia.org/wiki/" + title.replace(' ', '_'))
        }
    }

    private fun summary(title: String): Article? {
        val o = plainGet(
            "https://en.wikipedia.org/api/rest_v1/page/summary/" +
                title.replace(' ', '_').replace("/", "%2F")
        ) as? JsonObject ?: return null

        fun s(obj: JsonObject?, k: String) = (obj?.get(k) as? JsonPrimitive)?.contentOrNull
        val resolved = s(o, "title") ?: return null
        if (s(o, "type")?.contains("not_found") == true) return null

        return Article(
            title = resolved,
            description = s(o, "description"),
            extract = s(o, "extract"),
            url = ((o["content_urls"] as? JsonObject)?.get("desktop") as? JsonObject)
                ?.let { s(it, "page") },
            thumbnail = (o["thumbnail"] as? JsonObject)?.let { s(it, "source") },
        )
    }

    // ── Transport ────────────────────────────────────────────────

    private suspend fun throttled(url: String): JsonElement? {
        throttle.withLock {
            val wait = MIN_INTERVAL_MS - (System.currentTimeMillis() - lastCall)
            if (wait > 0) kotlinx.coroutines.delay(wait)
            lastCall = System.currentTimeMillis()
        }
        return plainGet(url)
    }

    /** Wikipedia has no comparable limit at this volume, so it skips the throttle. */
    private fun plainGet(url: String): JsonElement? = runCatching {
        val req = Request.Builder().url(url)
            .header("User-Agent", UA)
            .header("Accept", "application/json")
            .build()
        client.newCall(req).execute().use { resp ->
            if (!resp.isSuccessful) return null
            Json.parseToJsonElement(resp.body?.string().orEmpty())
        }
    }.getOrNull()
}

File: backend/modules/geo_intel.py
"""
Geo intelligence — place identification from coordinates, using only free,
keyless services.

This is the honest substitute for "landmark recognition". Visual landmark
recognition needs a paid vision API; this module instead resolves the GPS
coordinates that many photos already carry, which — when they are present — is
*more* reliable than recognising a building from pixels. Big Ben resolves to
"Big Ben, Bridge Street, Westminster" from coordinates alone, with the Wikidata
id and the eight nearest notable places attached.

The trade is stated plainly in the output: `requires_gps` is always true. No GPS
means no result, and the module says so rather than guessing from image content.

Sources, all keyless and all with usage policies this module respects:
  • Nominatim (OpenStreetMap)  — reverse geocoding
  • Wikipedia REST + Action API — article summary and geosearch

Nominatim's policy requires a genuine identifying User-Agent and at most one
request per second from a single source. Both are enforced below. Exceeding it
gets an IP banned, which would take the feature down for every user, so the
throttle is not optional politeness.
"""

from __future__ import annotations

import threading
import time
from typing import Any, Optional

import requests

NOMINATIM_URL = "https://nominatim.openstreetmap.org/reverse"
WIKI_REST = "https://en.wikipedia.org/api/rest_v1/page/summary/"
WIKI_API = "https://en.wikipedia.org/w/api.php"

# Nominatim requires contact details in the agent string. A generic browser UA
# is a policy violation and gets blocked.
_UA = "MyRecon-OSINT/1.0 (https://myrecon.xyz; abuse@myrecon.xyz)"
_HEADERS = {"User-Agent": _UA, "Accept": "application/json"}
_TIMEOUT = 12

# One request per second, enforced across threads. Gunicorn runs 8 threads per
# worker, so a per-call sleep would not be enough on its own.
_MIN_INTERVAL = 1.05
_rate_lock = threading.Lock()
_last_call = 0.0


def _throttled_get(url: str, params: dict) -> Optional[dict]:
    """GET with the shared Nominatim rate limit applied. Returns None on any
    failure — geo enrichment is additive, so it must never break a scan."""
    global _last_call
    with _rate_lock:
        wait = _MIN_INTERVAL - (time.monotonic() - _last_call)
        if wait > 0:
            time.sleep(wait)
        _last_call = time.monotonic()
    try:
        resp = requests.get(url, params=params, headers=_HEADERS, timeout=_TIMEOUT)
        if resp.status_code != 200:
            return None
        return resp.json()
    except (requests.RequestException, ValueError):
        return None


def _plain_get(url: str, params: Optional[dict] = None) -> Optional[dict]:
    """Wikipedia has no comparable rate limit for this volume, so it skips the
    Nominatim throttle and does not delay the request."""
    try:
        resp = requests.get(url, params=params, headers=_HEADERS, timeout=_TIMEOUT)
        if resp.status_code != 200:
            return None
        return resp.json()
    except (requests.RequestException, ValueError):
        return None


def _valid_coords(lat: Any, lon: Any) -> bool:
    try:
        la, lo = float(lat), float(lon)
    except (TypeError, ValueError):
        return False
    # Reject the null island exactly: 0,0 is overwhelmingly a zeroed GPS field
    # rather than a genuine position in the Gulf of Guinea.
    if la == 0 and lo == 0:
        return False
    return -90 <= la <= 90 and -180 <= lo <= 180


# ──────────────────────────────────────────────────────────────────────
#  Reverse geocoding
# ──────────────────────────────────────────────────────────────────────

def reverse_geocode(lat: float, lon: float) -> dict:
    """
    Resolve coordinates to a named place and administrative hierarchy.

    Nominatim's `zoom` is left at its default so it returns the most specific
    named feature it has — which is what surfaces "Big Ben" rather than merely
    "Westminster".
    """
    out: dict[str, Any] = {"found": False, "source": "nominatim"}
    if not _valid_coords(lat, lon):
        out["error"] = "Coordinates are missing or invalid."
        return out

    data = _throttled_get(NOMINATIM_URL, {
        "lat": lat, "lon": lon, "format": "jsonv2",
        "addressdetails": 1, "extratags": 1,
    })
    if not data or "error" in data:
        out["error"] = "Reverse geocoding was unavailable."
        return out

    addr = data.get("address") or {}
    extra = data.get("extratags") or {}
    # Nominatim varies which key holds the settlement depending on country, so
    # fall through the plausible ones in decreasing specificity.
    settlement = (addr.get("city") or addr.get("town") or addr.get("village")
                  or addr.get("municipality") or addr.get("county"))

    out.update({
        "found": True,
        "name": data.get("name") or None,
        "display_name": data.get("display_name"),
        "category": data.get("category"),
        "type": data.get("type"),
        "address": {
            "road": addr.get("road"),
            "suburb": addr.get("suburb") or addr.get("neighbourhood"),
            "city": settlement,
            "state": addr.get("state"),
            "postcode": addr.get("postcode"),
            "country": addr.get("country"),
            "country_code": (addr.get("country_code") or "").upper() or None,
        },
        "wikidata": extra.get("wikidata"),
        "wikipedia": extra.get("wikipedia"),
        "osm": {
            "type": data.get("osm_type"),
            "id": data.get("osm_id"),
            "url": (f"https://www.openstreetmap.org/{data.get('osm_type')}/{data.get('osm_id')}"
                    if data.get("osm_type") and data.get("osm_id") else None),
        },
    })
    return out


# ──────────────────────────────────────────────────────────────────────
#  Wikipedia
# ──────────────────────────────────────────────────────────────────────

def nearby_places(lat: float, lon: float, radius_m: int = 1000, limit: int = 8) -> list[dict]:
    """Notable places within radius, nearest first. Radius is capped at
    Wikipedia's documented maximum of 10 km."""
    if not _valid_coords(lat, lon):
        return []
    data = _plain_get(WIKI_API, {
        "action": "query", "list": "geosearch",
        "gscoord": f"{lat}|{lon}",
        "gsradius": max(10, min(int(radius_m), 10000)),
        "gslimit": max(1, min(int(limit), 50)),
        "format": "json",
    })
    if not data:
        return []
    return [{
        "title": item.get("title"),
        "distance_m": round(item.get("dist", 0), 1),
        "lat": item.get("lat"),
        "lon": item.get("lon"),
        "url": "https://en.wikipedia.org/wiki/" + str(item.get("title", "")).replace(" ", "_"),
    } for item in (data.get("query") or {}).get("geosearch") or []]


def article_summary(title: str) -> Optional[dict]:
    """Lead extract and thumbnail for a Wikipedia article."""
    if not title:
        return None
    data = _plain_get(WIKI_REST + str(title).replace(" ", "_"))
    if not data or data.get("type") == "https://mediawiki.org/wiki/HyperSwitch/errors/not_found":
        return None
    return {
        "title": data.get("title"),
        "description": data.get("description"),
        "extract": data.get("extract"),
        "url": ((data.get("content_urls") or {}).get("desktop") or {}).get("page"),
        "thumbnail": (data.get("thumbnail") or {}).get("source"),
        "coordinates": data.get("coordinates"),
    }


# ──────────────────────────────────────────────────────────────────────
#  Orchestration
# ──────────────────────────────────────────────────────────────────────

def _confidence(place: dict, article: Optional[dict], nearest_m: Optional[float]) -> dict:
    """
    How firmly the coordinates identify a *named* place.

    The coordinates themselves are as accurate as the device that recorded
    them; what varies is whether they land on something identifiable. A named
    OSM feature with a matching Wikipedia article a few metres away is a
    confident identification. A bare road in open country is not.
    """
    factors: list[str] = []
    score = 0
    if place.get("found"):
        score += 30
        factors.append("coordinates resolved to an address")
    if place.get("name"):
        score += 25
        factors.append("coordinates fall on a named feature")
    if place.get("wikidata"):
        score += 20
        factors.append("feature has a Wikidata entity")
    if article:
        score += 15
        factors.append("matching Wikipedia article found")
    if nearest_m is not None and nearest_m <= 50:
        score += 10
        factors.append(f"notable place within {nearest_m:.0f} m")

    score = min(100, score)
    label = "high" if score >= 70 else "medium" if score >= 40 else "low"
    return {"score": score, "band": label, "factors": factors}


def investigate_location(lat: float, lon: float, *, radius_m: int = 1000) -> dict:
    """
    Full location report for a coordinate pair.

    Intended to be fed straight from image_forensics' GPS block. Returns a
    `found: False` result rather than raising when there is nothing to report,
    so callers can include it unconditionally.
    """
    if not _valid_coords(lat, lon):
        return {
            "found": False,
            "requires_gps": True,
            "error": "No usable GPS coordinates.",
            "provenance": {"sources": [], "keyless": True},
        }

    place = reverse_geocode(lat, lon)
    nearby = nearby_places(lat, lon, radius_m=radius_m)

    # Prefer the article Nominatim points at; fall back to the closest notable
    # place, which is usually the same feature under its common name.
    article = None
    wiki_tag = place.get("wikipedia")
    if wiki_tag:
        # Format is "en:Article Title".
        article = article_summary(wiki_tag.split(":", 1)[-1])
    if not article and place.get("name"):
        article = article_summary(place["name"])
    if not article and nearby:
        article = article_summary(nearby[0]["title"])

    nearest = nearby[0]["distance_m"] if nearby else None

    return {
        "found": bool(place.get("found")),
        "requires_gps": True,
        "coordinates": {
            "latitude": round(float(lat), 6),
            "longitude": round(float(lon), 6),
            "maps": {
                "openstreetmap": f"https://www.openstreetmap.org/?mlat={lat}&mlon={lon}#map=17/{lat}/{lon}",
            },
        },
        "place": place,
        "article": article,
        "nearby": nearby,
        "confidence": _confidence(place, article, nearest),
        "provenance": {
            "sources": [
                {"provider": "nominatim", "url": "https://nominatim.openstreetmap.org/",
                 "licence": "ODbL (OpenStreetMap contributors)", "keyless": True},
                {"provider": "wikipedia", "url": "https://en.wikipedia.org/",
                 "licence": "CC BY-SA", "keyless": True},
            ],
            "keyless": True,
            "note": "Derived from embedded GPS, not from image content. "
                    "No visual landmark recognition is performed.",
        },
    }
